dali/python/nvidia/dali/plugin/mxnet.py

```diff
@@ -423,3 +459,224 @@ def __init__(self,
+            # Copy data from DALI Tensors to MXNet NDArrays
+            for j, output_el in enumerate(output_elements):
+                if self._outputs_types is None or self._outputs_types[j] == DALIGluonIterator.DENSE_TAG:
+                    ndarray = batch[j].resize(shapes[j])
+                    feed_ndarray(output_el, ndarray)
+                else:
+                    for sample_idx in range(self.batch_size):
+                        ndarray = batch[j][sample_idx].resize(shapes[j][sample_idx])
+                        feed_ndarray(output_el[sample_idx], ndarray)
+
+        batches = [[([sample.array for sample in output_el] if isinstance(output_el,list) else output_el.array)
+                    for output_el in batch]
+                   for batch in self._data_batches]
+
+        for p in self._pipes:
+            with p._check_api_type_scope(types.PipelineAPIType.ITERATOR):
+                p.release_outputs()
+                p.schedule_run()
+
+        self._counter += self._num_gpus * self.batch_size
+        if (not self._fill_last_batch) and (self._counter > self._size) and self._size > 0:
+            # First calculate how much data is required to return exactly self._size entries.
+            diff = self._num_gpus * self.batch_size - (self._counter - self._size)
+            # Figure out how many GPUs to grab from.
+            numGPUs_tograb = int(np.ceil(diff/self.batch_size))
+            # Figure out how many results to grab from the last GPU (as a fractional GPU batch may be required to
+            # bring us right up to self._size).
+            mod_diff = diff % self.batch_size
+            data_fromlastGPU = mod_diff if mod_diff else self.batch_size
+
+            # Grab the relevant data.
+            # 1) Grab everything from the relevant GPUs.
+            # 2) Grab the right data from the last GPU.
+            # 3) Append data together correctly and return.
+            output = batches[0:numGPUs_tograb]
+            output[-1] = output[-1].copy()
+            for element_idx in range(len(output[-1])):
+                output[-1][element_idx] = output[-1][element_idx][0:data_fromlastGPU]
+            return output
+
+        return batches
+
+    def _create_data_batch(self, output_elements, shapes, device_id):
+        mx_gpu_device = mx.gpu(device_id)
+        mx_cpu_device = mx.cpu(0)
+        from nvidia.dali.backend import TensorGPU
+        new_batch = []
+        for j, output_el in enumerate(output_elements):
+            first_t = output_el if self._outputs_types is None or self._outputs_types[j] == DALIGluonIterator.DENSE_TAG else output_el[0]
+            dtype = np.dtype(first_t.dtype())
+            device = mx_gpu_device if type(first_t) is TensorGPU else mx_cpu_device
+            if self._outputs_types is None or self._outputs_types[j] == DALIGluonIterator.DENSE_TAG:
+                new_batch.append(SmartArray(mx.nd.zeros(shapes[j], device, dtype=dtype)))
```

Comment:
Why not always create the mx.nd.zeros in the `SmartArray`?

Reply:
To keep SmartArray more flexible? I don't think it really matters and it reduces the number of arguments to pass to SmartArray ctor